text/0309-weak-cross-stack-references.md

```diff
@@ -0,0 +1,194 @@
+
+- In the first deployment, call [stack.exportValue](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.Stack.html#exportwbrvalueexportedvalue-options) on the referenced value in the producing stack, while removing the reference from the consuming stack.
+- In the second deployment, you can remove or replace the resource.
+
+### Weak cross-stack references
+
+If you want to avoid having to do a two-phase deployment when deleting or replacing certain resources, and you are confident that the consumer will not suffer negative impact if the resource is removed, you can force the reference to be a **weak reference**. Be sure that your consuming stack will continue to function properly though: for example, if you originally created an IAM Role in one stack and pass it to a Lambda Function in a different stack, and you make a change that cause the IAM Role to be replaced, there will be a period of time in between the deployments of the producer and consumer stacks in which the Lambda executions will fail.
+
+A common use case in which weak references are useful is **monitoring dashboards**: many people like setting up their dashboards in a separate stack and having it reference resources in other stacks in the same application; construct libraries like `cdk-watchful` make it very easy to fill a dashboard with graphs for various resources in the application. This requires transporting resource names (for example, a dashboard metric widget will contain `{ "Dimensions: { "FunctionName": { "Fn::ImportValue": "OtherStack-SomeLambda12345" } } }`)... but because of the behavior of strong references this effectively locks the `SomeLambda` construct into existence. It cannot be removed or replaced, because the dashboard reference would not be valid anymore. In cases like this, the friction is not worth the value it adds: there will be no negative downstream impact if the metrics stream the dashboard is observing stops emitting metrics, and most likely if the resource were to be removed, the corresponding graph would just be removed in the same deployment.
+
+To turn a strong reference into a weak reference, call `Reference.weakStringReference(...)` (or `weakListReference`, or `weakNumberReference`). For example, the following references a bucket name weakly:
+
+```ts
+const bucket = new s3.Bucket(this, 'MyBucket');
+
+// ...
+
+const fn = new lambda.Function(this, 'MyFunction', {
+  // ...
+
+  environment: {
+    BUCKET_NAME: Reference.weakStringReference(bucket.bucketName),
+  },
+});
+```
+
+`Reference.weakStringReference()` only has an effect if the reference is to a construct defined in a different stack: it will use SSM Parameter Store to transport the value between stacks. If the reference is to a construct in the same stack, nothing happens (do be aware that in the example above, if `Bucket` gets renamed, there is a certain time window in which the Lambda Function may continue to be executed with the old bucket name!).
```

Comment:
I'm having a rough time reading 

> only has an effect if the reference is to a construct defined in a different stack

mind rewording it maybe like

> the weak reference works only if its construct is defined in a different stack

?

Reply:
Should the time-window of doom warning get its own section with a couple of examples or do you think it's better to repeat it throughout the doc?